exporter/datadogexporter/translate_traces.go

```diff
@@ -0,0 +1,459 @@
+		Traces:       []*pb.APITrace{},
+		Transactions: []*pb.Span{},
+	}
+
+	if resource.IsNil() && ilss.Len() == 0 {
+		return payload, nil
+	}
+
+	resourceServiceName, datadogTags := resourceToDatadogServiceNameAndAttributeMap(resource)
+
+	// specification states that the resource level deployment.environment should be used for passing env, so defer to that
+	// https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/resource/semantic_conventions/deployment_environment.md#deployment
+	if resourceEnv, ok := datadogTags[deploymentEnv]; ok {
+		payload.Env = resourceEnv
+	}
+
+	apiTraces := map[uint64]*pb.APITrace{}
+
+	for i := 0; i < ilss.Len(); i++ {
+		ils := ilss.At(i)
+		if ils.IsNil() {
+			continue
+		}
+		extractInstrumentationLibraryTags(ils.InstrumentationLibrary(), datadogTags)
+		spans := ils.Spans()
+		for j := 0; j < spans.Len(); j++ {
+			span, err := spanToDatadogSpan(spans.At(j), resourceServiceName, datadogTags, cfg)
+
+			if err != nil {
+				return payload, err
+			}
+
+			var apiTrace *pb.APITrace
+			var ok bool
+
+			if apiTrace, ok = apiTraces[span.TraceID]; !ok {
+				// we default these values to 0 and then calculate the appropriate StartTime
+				// and EndTime within addToAPITrace()
+				apiTrace = &pb.APITrace{
+					TraceID:   span.TraceID,
+					Spans:     []*pb.Span{},
+					StartTime: 0,
+					EndTime:   0,
+				}
+				apiTraces[apiTrace.TraceID] = apiTrace
+			}
+
+			addToAPITrace(apiTrace, span)
+		}
+	}
+
+	for _, apiTrace := range apiTraces {
+		top := GetAnalyzedSpans(apiTrace.Spans)
+		ComputeSublayerMetrics(apiTrace.Spans)
+		payload.Transactions = append(payload.Transactions, top...)
+		payload.Traces = append(payload.Traces, apiTrace)
+	}
+
+	return payload, nil
+}
+
+// convertSpan takes an internal span representation and returns a Datadog span.
+func spanToDatadogSpan(s pdata.Span,
+	serviceName string,
+	datadogTags map[string]string,
+	cfg *Config,
+) (*pb.Span, error) {
+	// otel specification resource service.name takes precedence
+	// and configuration DD_ENV as fallback if it exists
+	if serviceName == "" && cfg.Service != "" {
+		serviceName = cfg.Service
+	}
+
+	version := cfg.Version
+	tags := aggregateSpanTags(s, datadogTags)
+
+	// if no version tag exists, set it if provided via config
+	if version != "" {
+		if tagVersion := tags[versionTag]; tagVersion == "" {
+			tags[versionTag] = version
+		}
+	}
+
+	// get tracestate as just a general tag
+	if len(s.TraceState()) > 0 {
+		tags[tracetranslator.TagW3CTraceState] = string(s.TraceState())
+	}
+
+	// get start/end time to calc duration
+	startTime := s.StartTime()
+	endTime := s.EndTime()
+	duration := int64(endTime) - int64(startTime)
+
+	// it's possible end time is unset, so default to 0 rather than using a negative number
+	if s.EndTime() == 0 {
+		duration = 0
+	}
+
+	datadogType := spanKindToDatadogType(s.Kind())
+
+	span := &pb.Span{
+		TraceID:  decodeAPMId(hex.EncodeToString(s.TraceID().Bytes()[:])),
```

Comment:
updated this to move the `hex.encodeToString` bit within the helper func. i prefer to leave it as accepting a byte array rather than a specific span data field, i think it makes testing a bit easier.